bottomless2/src/scheduler.rs

```diff
@@ -13,44 +30,342 @@ use crate::{
+
+        tracing::debug!(job_id = id, "job registered");
+
+        // if there is a priority for this namespace already, it must be higher than ours, because
+        // it was registered earlier
+        if !requests.in_flight && self.queue.get_priority(&name).is_none() {
+            tracing::debug!(job_id = id, "job queued");
+            self.queue.push(name, Reverse(id));
+        }
     }
 
     /// Get the next job to be executed. Gather as much work as possible from the next namespace to
     /// be scheduled, and returns description of the job to be performed. No other job for this
     /// namespace will be scheduled, until the `JobResult` is reported
+    #[tracing::instrument(skip_all)]
     pub fn schedule(&mut self) -> Option<Job<S>> {
-        todo!()
+        let (name, _) = self.queue.pop()?;
+        let requests = self
+            .requests
+            .get_mut(&name)
+            .expect("work scheduled but not requests?");
+        requests.in_flight = true;
+        let request = requests.requests.pop_front().unwrap();
+
+        tracing::debug!(job_id = request.id, "scheduled job");
+
+        let job = Job {
+            storage: self.storage.clone(),
+            request,
+        };
+
+        Some(job)
     }
 
     /// Report the job result to the scheduler. If the job result was a success, the request as
     /// removed from the queue, else, the job is rescheduled
-    pub fn report(&mut self, _job: JobResult<S>) {
+    #[tracing::instrument(skip_all, fields(req_id = result.job.request.id))]
+    pub async fn report(&mut self, result: JobResult<S>) {
         // re-schedule, or report new max durable frame_no for segment
-        todo!()
+        let name = result.job.request.request.namespace.clone();
+        let requests = self
+            .requests
+            .get_mut(&name)
+            .expect("request slot must exist");
+
+        requests.in_flight = false;
+
+        match result.result {
+            Ok(durable_index) => {
+                tracing::debug!("job success registered");
+                if self
+                    .durable_notifier
+                    .send((name.clone(), durable_index))
+                    .await
+                    .is_err()
+                {
+                    tracing::warn!("durability notifier was closed, proceeding anyway");
+                }
+            }
+            Err(e) => {
+                tracing::error!("error processing request, re-enqueuing");
+                // put it back at the front of the queue
+                requests.requests.push_front(result.job.request);
+            }
+        }
+
+        assert!(
+            self.queue.get_priority(&name).is_none(),
+            "there should be not enqueued jobs at this point"
+        );
+
+        if !requests.requests.is_empty() {
+            let first_id = requests.requests.front().unwrap().id;
+            self.queue.push(name, Reverse(first_id));
```

Comment:
feels a little odd to me that report also can enqueue the next job?

Reply:
it doesn't enqueue the next job, it just make that namespace available for scheduling again. The next job will be popped from he queue on the next call to `schedule`

Reply:
the idea is that you can't schedule work for a namespace that already has in-flight work. So when you schedule a namespace, you pop it from the queue, take some work from it to create a job and return the job. As long as the job is not reported, no more work for that namespace can be scheduled. When the job is reported, we put it back in the queue so that more work can be performed on it, whenever it's that namespace turn to make progress.